addons/account/migrations/13.0.1.1/post-migration.py

```diff
@@ -238,24 +238,63 @@ def migration_invoice_moves(env):
             AND ((ail.product_id IS NULL AND aml.product_id IS NULL) OR ail.product_id = aml.product_id)
             AND ((ail.uom_id IS NULL AND aml.product_uom_id IS NULL) OR ail.uom_id = aml.product_uom_id)
         """
-    # Try first with a stricter criteria for matching invoice lines with account move lines
-    openupgrade.logged_query(
-        env.cr, query + minimal_where + """
-            AND ail.account_id = aml.account_id
-            AND ai.commercial_partner_id = aml.partner_id
-            AND ((ail.account_analytic_id IS NULL AND aml.analytic_account_id IS NULL)
-                OR ail.account_analytic_id = aml.analytic_account_id)
-        RETURNING aml.id""",
-    )
-    aml_ids = tuple(x[0] for x in env.cr.fetchall())
-    # Try now with a more relaxed criteria, as it's possible that users change some data on amls
-    openupgrade.logged_query(
-        env.cr, query + minimal_where + """
-            AND aml.old_invoice_line_id IS NULL
-            AND rc.anglo_saxon_accounting IS DISTINCT FROM TRUE
-        RETURNING aml.id""",
-    )
-    aml_ids += tuple(x[0] for x in env.cr.fetchall())
+    aml_ids = tuple()
+    stricter = True
+    # Doing loop to fix where two or more move lines points to the same old invoice line
+    while True:
+        if stricter:
+            # Try first with a stricter criteria for matching invoice lines with account move lines
+            openupgrade.logged_query(
+                env.cr, query + """
+                LEFT JOIN account_move_line aml2 ON aml2.old_invoice_line_id = ail.id
+                """ + minimal_where + """
+                    AND ail.account_id = aml.account_id
+                    AND ai.commercial_partner_id = aml.partner_id
+                    AND ((ail.account_analytic_id IS NULL AND aml.analytic_account_id IS NULL)
+                        OR ail.account_analytic_id = aml.analytic_account_id)
+                    AND aml.old_invoice_line_id IS NULL AND aml2.id IS NULL
+                RETURNING aml.id""",
+            )
+            aml_ids += tuple(x[0] for x in env.cr.fetchall())
+            openupgrade.logged_query(
+                env.cr, """
+                UPDATE account_move_line aml
+                SET old_invoice_line_id = NULL
+                FROM (SELECT id, rank() OVER (PARTITION BY old_invoice_line_id ORDER BY id) repeated_invoice_line
```

Comment:
You want to avoid to execute this twice per loop just for performance. But I am more interested in assure that there is not a wrong matched invoice line. If we execute the two matching queries consecutively, it could happen that the one of lesser criteria matches one wrongly.

Reply:
I don't think so with the `LEFT JOIN` with already matched lines. Take into account that this case is minority and I'm not facing it, so for being accepted, this should be optimized the maximum.

Reply:
changes applied. I hope now it fits your taste :)